### PR TITLE
Updates to step-10

### LINKED ISSUE #9341: [Step-10] Remove long double from example step and improve discussion of round-off errors
In Step-10, we have

>Since we would like to compute the difference between two numbers which are quite accurate, with the accuracy of the computed approximation to π being in the range of the number of digits which a double variable can hold, we rather declare the reference value as a long double and give it a number of extra digits:
`const long double pi = 3.141592653589793238462643L;`

From my fairly rudimentary understanding of the C++ language, this may not be the best practice? The number of precision bits of `long double` strongly depends on a) compiler, b) OS and c) machine architecture.

Perhaps a better idea is to replace it with a predictable size of `pi` using a suitable multi-precision library. Since a standard Deal.II installation already bundles boost, do the developers think it is worth replacing this with the facilities provided by [`boost.multiprecision`](https://www.boost.org/doc/libs/1_72_0/libs/multiprecision/doc/html/index.html) for example?

### PR DESCRIPTION
**Pull request by bangerth**
step-10 uses `long double`, and we decided in #9341 that this is silly. This patch fixes it and updates the results section. I added a bit of discussion as mentioned in #9341 about summations etc, but I don't actually think that summation is the limiting factor because we almost certainly add numbers of roughly the same size here.

Fixes #9341.

/rebuild

### CODE REVIEW

File: examples/step-10/doc/results.dox [hunk 1]
```diff
@@ -217,3 +226,97 @@ convergence. (This effect is also discussed in detail in the following
...
+two. (Of course, in all cases, we have computed $\pi$ with more
+accuracy than any engineer would ever want to know.)
+
+What explains this unpredictability? In general, round-off errors can
+be thought of as random that add up in ways that are not worth thinking
+too much about; we should therefore always treat any accuracy beyond, say,
+the thirteenth digits as suspect. Thus, it is probably not worth spending
+too much time on wondering why we get different winners and losers in the
+data type exchange from `double` and `long double`. The accuracy of the
+results is also largely not determine by the precision of the data type
```
Comment (drwells) [inline examples/step-10/doc/results.dox:305]:
```suggestion
results is also largely not determined by the precision of the data type
```

File: examples/step-10/doc/results.dox [hunk 2]
```diff
@@ -217,3 +226,97 @@ convergence. (This effect is also discussed in detail in the following
...
+ 5120 3.1415926535897944 1.3798e-15 -0.00
+@endcode
+Here, using `double` precision is more accurate by about a factor of
+two. (Of course, in all cases, we have computed $\pi$ with more
+accuracy than any engineer would ever want to know.)
+
+What explains this unpredictability? In general, round-off errors can
+be thought of as random that add up in ways that are not worth thinking
+too much about; we should therefore always treat any accuracy beyond, say,
+the thirteenth digits as suspect. Thus, it is probably not worth spending
```
Comment (drwells) [inline examples/step-10/doc/results.dox:302]:
```suggestion
thirteen digits as suspect. Thus, it is probably not worth spending
```

File: examples/step-10/doc/results.dox [hunk 3]
```diff
@@ -217,3 +226,97 @@ convergence. (This effect is also discussed in detail in the following
...
+too much time on wondering why we get different winners and losers in the
+data type exchange from `double` and `long double`. The accuracy of the
+results is also largely not determine by the precision of the data type
+in which we accumulate each cell's (or face's) contributions, but the
+accuracy of what deal.II gives us via FEValues::JxW() and FEFaceValues::JxW(),
+which always uses `double` precision and which we cannot directly affect.
+
+But there are cases where one can do something about the precision, and it
+is worth at least mentioning the name of the most well-known algorithm in
+this algorithm. Specically, what we are doing when we add contributions into
```
Comment (drwells) [inline examples/step-10/doc/results.dox:312]:
```suggestion
this area. Specifically, what we are doing when we add contributions into
```

File: examples/step-10/step-10.cc [hunk 4]
```diff
@@ -366,13 +345,13 @@ namespace Step10
...
+                      perimeter += fe_face_values.JxW(i);
                   }
             // Then store the evaluated values in the table...
-            table.add_value("eval.pi", static_cast<double>(perimeter / 2.0L));
-            table.add_value(
-              "error", static_cast<double>(std::fabs(perimeter / 2.0L - pi)));
+            table.add_value("eval.pi", static_cast<double>(perimeter / 2.0));
+            table.add_value("error",
+                            static_cast<double>(
+                              std::fabs(perimeter / 2.0 - numbers::PI)));
```
Comment (drwells) [inline examples/step-10/step-10.cc:354]: do we still need this static cast?

**Review by drwells — CHANGES_REQUESTED** [review]

File: examples/step-10/doc/results.dox [hunk 5]
```diff
@@ -217,3 +226,97 @@ convergence. (This effect is also discussed in detail in the following
...
+  320 3.1415926535897576 3.5705e-14  7.93
+ 1280 3.1415926535897918 1.3785e-15  4.70
+ 5120 3.1415926535897944 1.3798e-15 -0.00
+@endcode
+Here, using `double` precision is more accurate by about a factor of
+two. (Of course, in all cases, we have computed $\pi$ with more
+accuracy than any engineer would ever want to know.)
+
+What explains this unpredictability? In general, round-off errors can
+be thought of as random that add up in ways that are not worth thinking
```
Comment (jppelteret) [inline examples/step-10/doc/results.dox:300]: >as random

Perhaps this is incomplete -- maybe `as random errors`? Or `as random, and add up in ways...`

File: examples/step-10/step-10.cc (hunk 4 above)
Comment (bangerth) [reply examples/step-10/step-10.cc:354]: No. Good eyes!

File: examples/step-10/doc/results.dox (hunk 5 above)
Comment (bangerth) [reply examples/step-10/doc/results.dox:300]: Yes, the latter is what I meant.

**Comment by bangerth** [comment]
So fixed.

**Review by drwells — APPROVED** [review]

**Review by jppelteret — APPROVED** [review]

**Pull request merged**